doc: update conditions, add "deno" and "types"

guybedford opened the pull request:
This adds two new conditions to the endorsed "exports" conditions list in the Node.js documentation.

1. "deno" - as of https://github.com/denoland/deno/pull/12424/files#diff-0523983fa24098bdaddcdb01d5f3771335b9e564df0c81d4b79101e75a108718R34, Deno now supports matching the `"deno"` condition when resolving using its Node.js compatibility resolver, allowing packages to define variants specifically for Deno. This makes the text in the Node.js documentation incorrect as since now the implementation exists we can correspondingly endorse its usage.
2. "types" - Angular have adopted this condition in the "exports" field as described in https://rc.angular.io/guide/angular-package-format#exports, explicitly calling out this pattern for typing can help move typing forward for the "exports" field.

It also reorders conditions by specificity and clarifies some of the definitions including:

* Browser is now defined strictly to be browser
* Development is strictly defined to imply debugging context

These could possibly be split out into separate PRs for further discussion if there are any objections to any of the specific conditions listed in this PR.

@nodejs/modules

guybedford commented:
I've amended this PR to also include a `"types"` condition //cc @weswigham would be great to get your feedback on this as well.

Trott commented on doc/api/packages.md:
```diff
@@ -612,16 +612,22 @@ The following condition definitions are currently endorsed by Node.js:
-may be used, but while there remain no implementation or integration intent
+Platform specific conditions such as `"electron"`, or `"react-native"` may
+be used, but while there remain no implementation or integration intent
 from these platforms, the above are not explicitly endorsed by Node.js.
```
Out of scope for this PR perhaps, so feel free to ignore, but [our style guide](https://github.com/nodejs/node/blob/master/doc/guides/doc-style-guide.md) refers folks to the [Microsoft Writing Style Guide](https://docs.microsoft.com/en-us/style-guide/welcome/) for things not covered in our style guide. Microsoft style would be more like this:

```suggestion
from these platforms, we do not recommend them.
```

Refs: https://docs.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/r/recommend
Refs: https://docs.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/a/above

GeoffreyBooth commented on doc/api/packages.md:
```diff
@@ -612,17 +612,23 @@ The following condition definitions are currently endorsed by Node.js:
   entry point. _Must always be mutually exclusive with `"production"`._
 * `"production"` - can be used to define a production environment entry
   point. _Must always be mutually exclusive with `"development"`._
+* `"source"` - indicates the original .js source file without minification
+  or bundling optimizations, useful for development and debugging workflows.
```
Is there an expectation that this is runnable? Like could this point to TypeScript files, or JavaScript that uses unsupported/experimental features like decorators?

weswigham replied:
Personally, the sourcemaps associated with a file are always going to be how you locate "original sources" - they contain strictly more information than a simple export map field can express. I don't know what tooling uses such a condition.

targos replied:
I hope it can point to TypeScript files, which can be made runnable using ESM loaders or require extensions.

guybedford replied:
There is no reason that `"exports"` can't point to `.ts` files if targetting a platform that supports TypeScript (eg when using the `"deno"` condition). But the expectation for the `"source"` condition would typically be `.js` so the sources after running tsc but before bundling. Source maps are good and well, but if you've got a monorepo with `tsc` running, and the bundler optimization as a separate step it's a very useful workflow and a fast monorepo dev process. Many users don't yet know about `--enable-source-maps` and they do still have debuggability issues anyway. I can appreciate if deferring this condition makes sense for now though.

targos replied:
I feel like this condition is very specific to your use case / expectations, so it probably makes sense to discuss it separately from the other ones.

bmeck replied:
I'd agree, the rest seem to match up with using the whole loader pipeline for something, but this is explicitly asking loaders not to perform part of the pipeline if they would normally do so.

GeoffreyBooth replied:
As a naïve user I would expect this to be whatever the original sources were, like what `"module"` has traditionally pointed to. If this is supposed to point to runnable-but-unminified JavaScript, perhaps a better name would suggest that.

weswigham replied:
Isn't that _generally_ what the `development` condition would be used for?

guybedford replied:
@weswigham no, "development" is supposed to be a replacement for the `process.env.NODE_ENV === 'production'` condition for providing different behaviour between development and production environments for library consumers like React does.

If this is unclear perhaps we can clarify the current definition further.

@bmeck @targos agreed `"source"` might need some more thought for now, will remove it.

guybedford replied:
@weswigham this is a good point though - we could indicate that the "production" condition can be used for minification to try and encourage such a pattern. The question is really then - do we want to encourage such a pattern? My concern with it is that most applications will always have a dependency that is unminified, so the minification status of individual dependencies is largely irrelevant if just one implies needing to reminify the build anyway.

bmeck replied:
I have a slight preference that if the expectation is to be in production, we don't define an explicit condition for it and instead people should use guards against production like `"development"`. Having both worries me a little bit since it complicates the number of values in order to represent what is possible to do already (if a little more verbose for some trees)

guybedford replied:
I tend to feel the same way and that was why it's a `"source"` condition and not an `"optimized"` condition. I still feel it is independent to `"development"` though somewhat, in that a development build can still be optimized.

Point taken on the `"production"` condition, it could be worth considering deprecating this condition in due course even.

weswigham commented on doc/api/packages.md:
```diff
@@ -612,17 +612,23 @@ The following condition definitions are currently endorsed by Node.js:
+* `"source"` - indicates the original .js source file without minification
+  or bundling optimizations, useful for development and debugging workflows.
+* `"types"` - can be used by typing systems to resolve the typing file for
+  the given export, possible since the interface should be the same for all
```
Explicitly: you need to have nested conditional `types` entrypoints if your conditional `exports` entrypoints have differing shapes (eg, because one is a cjs implementation). Meaning a map like:
```json
{
  "exports": {
    "types": {
      "import": "./types/esm/index.d.mts",
      "default": "./types/cjs/index.d.ts"
    },
    "import": "./dist/esm/index.mjs",
    "default": "./dist/cjs/index.js"
  }
}
```



90% of the time it's better to just let your declaration files live alongside your js files than figure this stuff out yourself, though. There are honestly too many axis exports can reasonably vary on (import vs require, dev vs prod, types or not), and it turns out the "the order of matching conditions in the object is the order lookups are attempted in" is confusing in practice, so removing one of those axis is probably preferable when possible.

guybedford replied:
@weswigham point taken, although for large numbers of variations surely this results in a lot of duplicated typing files? The point of such a condition would be to avoid this. I certainly wouldn't recommend it just yet, and I know there are still the dual mode issues, but for ESM only packages with browser / development and Deno variations it seems like a way to avoid three copies of the typing files to me.

guybedford replied:
Alternatively, perhaps the `"source"` condition could be used for typing? As the basic tsc emit with side-by-side declarations, before the second-pass optimization?

aduh95 commented on doc/api/packages.md:
```diff
@@ -612,17 +612,23 @@ The following condition definitions are currently endorsed by Node.js:
   entry point. _Must always be mutually exclusive with `"production"`._
 * `"production"` - can be used to define a production environment entry
   point. _Must always be mutually exclusive with `"development"`._
+* `"source"` - indicates the original .js source file without minification
```
```suggestion
* `"source"` - indicates the original `.js` source file without minification
```

weswigham commented on doc/api/packages.md:
```diff
@@ -612,17 +612,23 @@ The following condition definitions are currently endorsed by Node.js:
   point. _Must always be mutually exclusive with `"development"`._
+* `"source"` - indicates the original .js source file without minification
+  or bundling optimizations, useful for development and debugging workflows.
+* `"types"` - can be used by typing systems to resolve the typing file for
```
Oh, this definitely needs to call out that `types` needs to come _before_ other conditions that will be active like `import`, `require`, or `default` - otherwise it won't be matched. We're seeing _extreme_ confusion about this in the community. It's really bad.

Potentially also worth noting is that TS supports conditions like `"types@>=4.3"` so you can semver-limit your type resolutions.

guybedford replied:
@weswigham is there a reference for the types versioning that we can link to here?

guybedford replied:
I've noted that `"types"` should always come first. We do generally note that conditions should be from most specific to least specific. Ways to better clarify this are welcome. Perhaps a diagram that clearly demonstrates with an example that it's just like a set of if conditions would clearly get this across....

GeoffreyBooth replied:
We could also list the conditions in order of precedence here in this document, rather than putting `"types"` last.

guybedford replied:
Good point - I've ordered all condition lists from most specific to least specific and added another note to each list about this.

guybedford commented:
I've removed the "source" condition for now, and also included a clarification that the "development" condition is primarily for debugging context use cases.

GeoffreyBooth commented on doc/api/packages.md:
```diff
@@ -486,8 +486,17 @@ For example, a package that wants to provide different ES module exports for
+  on native addons. This condition can be disabled via the
+  [`--no-addons` flag][].
+* `"node"` - matches for any Node.js environment. Can be a CommonJS or ES
+  module file. _This condition should always come after `"import"` or
+  `"require"`._
```
If it should always come after `import` or `require`, why is it listed before them here?

guybedford replied:
It's kind of a category question as specificity isn't a single vector in reality, we're kind of projecting it as one here though. I think this advice was given before nested conditionals were fully supported as well. In theory node is more specific so should come early by the general rule, but if using alongside require and import should then use nesting or precedence.

How about:

```suggestion
  module file. _should only be used when absolutely necessary, and in most cases it isn't. If also using the `"import"` or `"require"` dual mode conditions, they should be nested under this condition._
```

GeoffreyBooth replied:
Better, but I think you need to explain "when absolutely necessary" might be necessary.

guybedford replied:
I've updated the description to simply not call out this interaction. Previously when we had fewer conditions, it was the first example of a nested interaction, while now I think the principle of nesting and interactions of "require" and "import" likely applies to all conditions equally.

GeoffreyBooth commented on doc/api/packages.md:
```diff
@@ -495,8 +495,8 @@ specific to least specific as conditions should be defined:
 * `"node"` - matches for any Node.js environment. Can be a CommonJS or ES
-  module file. _This condition should always come after `"import"` or
-  `"require"`._
+  module file. _In most cases explictily calling out the Node.js platform is
+  not necessary._
```
C’mon, gimme just one case where it’s necessary 😄 

You don’t have to add it to the docs, but I’m just very curious now when this could be required.

guybedford replied:
JSPM Core is a good example, see eg https://github.com/jspm/jspm-core/blob/main/package.json#L44. Basically when Node.js really is the only intention (it directs the core modules through `export * from 'node:[mod]'; export { default } from 'node:[mod]'`). This condition comes more into its own when there are more runtimes than just Deno and Node as currently "deno" sufficies as a negative condition anyway in most cases for alternative non-browser platforms.

GeoffreyBooth replied:
Sounds like maybe we need a condition that's something like "server-side" as in, not browsers. Appropriate for either Node or Deno, maybe other command-line runtimes as well.

guybedford replied:
We have "not browser" and that is the "browser" / "default" condition. All not conditions are well defined. New platform definitions are harder to define in a really concrete way short of being specific platforms.

We did say that "browser" is any web standards API with a DOM, but now that Deno has Node.js compat, perhaps we should consider making "browser" really mean a web browser or web-browser like environment to help ensure this?

GeoffreyBooth replied:
For that matter, isn't Deno adding support for a DOM? And Node already has JSDom. The line is blurry.

guybedford replied:
Right which is why I'm wondering if maybe we should clarify browser means browser as opposed to trying to define some browser subset. This way not browser is clearer as well. Previously matching "browser" for Deno was a useful compat mechanism but now that we have "deno" and a compat mode for Deno, it seems unnecessary to treat Deno support as matching "browser" conditions (and similarly for other non-browser platforms).

guybedford replied:
I've gone ahead and updated the browser condition clarification here. Let me know how that seems.

bmeck replied:
For DataDog's `dd-trace` package, unfortunately CDNs/bundlers often ignore that it isn't meant to load on the browser and often generate some form of transformation if minor effort is made. For example, we are adding the explicit `node` condition to prevent things like https://cdn.skypack.dev/dd-trace from happening and getting support errors. It also would not work in `deno` or other server envs.

guybedford replied:
Marking this as resolved, but let me know if any wording updates or clarifications can be made.

ljharb commented on doc/api/packages.md:
```diff
@@ -486,8 +486,17 @@ For example, a package that wants to provide different ES module exports for
+  on native addons. This condition can be disabled via the
+  [`--no-addons` flag][].
+* `"node"` - matches for any Node.js environment. Can be a CommonJS or ES
+  module file. _In most cases explictily calling out the Node.js platform is
```
```suggestion
  module file. _In most cases explictly calling out the Node.js platform is
```

guybedford replied:
Well-spotted, although still had a typo 😂

```suggestion
  module file. _In most cases explicitly calling out the Node.js platform is
```

ljharb commented:
It seems really strange to me to explicitly endorse a private company’s product in node’s documentation.

guybedford commented:
> It seems really strange to me to explicitly endorse a private company’s product in node’s documentation.

We originally discussed having a conditions registry being maintained outside of Node.js, but then made the decision to maintain the list in the Node.js documentation for now.

No one has so far expressed any issue about listing `"deno"` here, but if there is further concern we could discuss alternatives.

ljharb commented:
The other conditions are all broad community standards, like browsers or types, or node things themselves.

I don’t think it’s really appropriate for node to claim a “registry” when node isn’t enforcing anything like that - it seems better for each tool to describe the conditions that tool supports.

guybedford commented:
@ljharb I personally feel quite strongly that especially given the risks of confusion we should be putting together more guidance not less with regards to usage of the "exports" field, and that maintaining a well-defined list of conditions most JS developers should know about is part of that effort.

Your argument here seems to be on the grounds of "appropriate", which seems quite ill defined. Can you try to explain more clearly what exactly you disagree with about Node.js mentioning the "deno" condition? IMO this isn't a case of companies competing, but platforms coordinating in the shared interest of the user.

ljharb commented:
I agree with you that those resources should exist, but it feels like the sort of thing the community should be deciding - node’s docs are authoritative and supposed to be primarily about how to use node.

There’s no way imo that “most” JS developers “should” know about a lot of the options for this list - they’re things that a subset of developers, and often a small subset, will want to know about. Package authors targeting a non-node platform generally will be familiar with that platform’s docs, as will tooling authors - who else needs to know about possible exports conditions?

GeoffreyBooth commented on doc/api/packages.md:
```diff
@@ -607,22 +608,28 @@ Runtimes or tools other than Node.js can use them at their discretion.
+* `"types"` - can be used by typing systems to resolve the typing file for
+  the given export, possible since the interface should be the same for all
+  variations. _This condition should always be included first._
+* `"deno"` - indicates a variation for the Deno platform.
```
Is there something in the Deno docs we could link to? Like any mention of Deno’s support for `"exports"`, if it has any, or its Node.js compatibility mode?

guybedford replied:
This is currently undocumented in Deno. Hence why I think it is important to highlight from a package compat perspective as it as much up to us to show how to make universal packages as anyone else.

ljharb replied:
Is Deno unwilling to document it? It seems very strange that "not deno" would be deciding interop norms for deno.

guybedford replied:
@ljharb Deno can and should document it, the point is that the dominant ecosystem can define the interop norms much better than the less popular platform alternatives.

There's been a lot of discussion I've yet to hear a strong argument for why we should _not_ include Deno to help users understand how conditions can be used to define universal packages. How does that materially cause any issues here?

GeoffreyBooth replied:
I think this is good to add, and there's nothing controversial about saying that the `deno` condition intends the Deno platform. I would expect them to do the same for us.

ljharb replied:
I think it's quite controversial for node to declare the interop story for deno. If deno's decided it, then they should document it first, and we should link to it - but if they haven't documented it yet, then "not deno" should not be the first party to do so.

mcollina replied:
I agree with @ljharb.

mhdawson replied:
From my understanding I also agree with @ljharb's last comment. Not that we should not document at some point but the first place being in the node repo does not sound right.

GeoffreyBooth replied:
We could submit a PR to Deno’s docs to suggest adding a mention of this. Then we would learn how they felt about it. If they merge in the PR, then we could add this reference here and link to their docs.

bartlomieju replied:
Hi folks! I did an update to the manual that explains this bit in more detail and links back to Node's docs: https://deno.land/manual@main/npm_nodejs/compatibility_mode Let me know what you think, thanks!

GeoffreyBooth commented on doc/api/packages.md:
```diff
@@ -607,22 +608,28 @@ Runtimes or tools other than Node.js can use them at their discretion.
 * `"development"` - can be used to define a development-only environment
-  entry point. _Must always be mutually exclusive with `"production"`._
+  entry point, for example to provide additional debugging context such as
+  better error message when running in a development mode. _Must always be
```
```suggestion
  better error messages when running in a development mode. _Must always be
```

GeoffreyBooth commented on doc/api/packages.md:
```diff
@@ -607,22 +608,28 @@ Runtimes or tools other than Node.js can use them at their discretion.
-Platform specific conditions such as `"deno"`, `"electron"`, or `"react-native"`
-may be used, but while there remain no implementation or integration intent
-from these platforms, the above are not explicitly endorsed by Node.js.
+Platform specific conditions such as `"electron"`, or `"react-native"` may
```
```suggestion
Platform-specific conditions such as `"electron"` or `"react-native"` may
```

guybedford replied:
Added the hyphen but not the comma change as this change was separately made previously per the docs styling linter I believe.

GeoffreyBooth commented on doc/api/packages.md:
```diff
@@ -607,22 +608,28 @@ Runtimes or tools other than Node.js can use them at their discretion.
-from these platforms, the above are not explicitly endorsed by Node.js.
+Platform specific conditions such as `"electron"`, or `"react-native"` may
+be used, but while there remain no implementation or integration intent
+from these platforms, we do not yet explicitly recommend them.
```
```suggestion
from these platforms, they are not explicitly endorsed by Node.js.
```

guybedford replied:
This was as suggested by @Trott in https://github.com/nodejs/node/pull/40708#discussion_r741622618, I'm open to a suggestion that takes into account the original feedback.

Trott replied:
If you wanted to make the case that "endorse" is sufficiently different from "recommend" that my original comment doesn't apply, I wouldn't argue the point.

GeoffreyBooth replied:
I just intended to restore the passive voice from the original version. I don't care re recommend vs endorse.

guybedford replied:
Ok I've landed the passive voice with recommended!

mcollina reviewed (changes requested):
I'm -1 in listing deno here without a @nodejs/tsc approval.

mhdawson commented:
Adding TSC agenda so that we make sure to discuss.

guybedford commented:
This section of the documentation was specifically created to list conditions in use in the JS ecosystem as a community resource. If there is resistance to Node.js maintaining this resource, then perhaps we can move it to its own repo instead.

If we do this, @mcollina @ljharb would you be against the Node.js docs linking to such a resource if there were a community maintained list of condition definitions repo?

ljharb commented:
I think it's totally appropriate to have an unofficial, non-node-affiliated community resource for this, and as long as it's described as such, the node docs certainly could link to it (as they do for a number of other unofficial things, like using apt to install node)

Trott commented on doc/api/packages.md:
```diff
@@ -607,22 +608,28 @@ Runtimes or tools other than Node.js can use them at their discretion.
-Platform specific conditions such as `"deno"`, `"electron"`, or `"react-native"`
-may be used, but while there remain no implementation or integration intent
-from these platforms, the above are not explicitly endorsed by Node.js.
+Platform-specific conditions such as `"electron"`, or `"react-native"` may
```
Optional micro-nit:

```suggestion
Platform-specific conditions such as `"electron"`, or `"react-native"` can
```

Refs: https://docs.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/c/can-may

Trott commented on doc/api/packages.md:
```diff
@@ -607,22 +608,28 @@ Runtimes or tools other than Node.js can use them at their discretion.
-from these platforms, the above are not explicitly endorsed by Node.js.
+Platform-specific conditions such as `"electron"`, or `"react-native"` may
+be used, but while there remain no implementation or integration intent
+from these platforms, they are not currently explicitly recommended.
```
Optional nit. I think this might be more clear:

```suggestion
from these platforms, we do not recommended them.
```

GeoffreyBooth replied:
In professional documentation, I expect the passive voice. It’s not clear who the “we” would be here.

Trott replied:
The use of "we recommend" is from the Microsoft Style Guide, which we've adopted so that we can keep out own local style guide nice and short. (It only needs a few parochial things like making sure we use _Node.js_ and not _NodeJS_.)

While they don't sound enthusiastic about it, they explicitly say "we recommend" is OK, and explicitly say to avoid passive voice "it is recommended". Ref: https://docs.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/r/recommend

The reasoning is probably the reasoning in their [Three voice principles](https://docs.microsoft.com/en-us/style-guide/brand-voice-above-all-simple-human#three-voice-principles):

> Our voice is:
> 
> * Warm and relaxed—We’re natural. Less formal, more grounded in real, everyday conversations. Occasionally, we’re fun. (We know when to celebrate.)

Trott replied:
To me, it's reasonably clear that "we" here refers to the Node.js project. However, if you want to be a bit more formal, you can remove the "recommend"/"endorse" stuff entirely and be imperative:

```suggestion
from these platforms, avoid them when possible.
```

...or something along those lines.

Trott replied:
All of that said, I'm OK if my nit here is ignored. While I'd _prefer_ one of my suggestions be adopted so we have consistent style in accordance with our chosen style guide etc., I'm definitely more interested in not burning out contributors on purely stylistic doc nits!

GeoffreyBooth replied:
The current version was passive, which is why it stuck out to me that this was proposing changing it. Perhaps the greater issue here is that this is very clearly an opinion; why should the user “avoid them when possible,” for example? Maybe we should take an extra sentence or two and explain why the user should do the thing a certain way, and then we don’t need make a recommendation or express an opinion. Something like:

```suggestion
Platform-specific conditions such as `"electron"`, or `"react-native"` may
be used when the target is unavoidably a single runtime; but maximum
compatibility with all potential consumers can be achieved by using broader
targets such as `"import"`.
```

guybedford replied:
Taking a step back, this entire section is actually redundant. We already note the application-specific non-listing in the requirements for custom definition listings. This section was originally written before we had other custom definitions in place, so was initiall more about giving examples of what conditions might be used for beyond the Node.js ones. Since this is no longer necessary and is adding more confusion I've gone ahead and removed it.

guybedford commented:
I've added a commit to clarify that the "conditions definitions" Node.js docs section is an entire documentation section dedicated to maintaining a list of conditions defintions of conditions not directly implemented in Node.js as a community coordination help.

In the process I removed some redundant text and also realised the `node-addons` condition description had been incorrectly placed into this section when that is a Node.js condition so does not belong in the definitions registry section.

To help further clarify this docs section, we could also consider renaming the title from "condition definitions" to "community definitions" to be clear that this is a coordination effort and not to do with documenting Node.js.

mcollina commented on doc/api/packages.md:
```diff
@@ -596,34 +603,30 @@ The `"import"`, `"require"`, `"node"`, `"node-addons"` and `"default"`
 
-Platform specific conditions such as `"deno"`, `"electron"`, or `"react-native"`
-may be used, but while there remain no implementation or integration intent
-from these platforms, the above are not explicitly endorsed by Node.js.
```
Are we losing "Electron" and "React Native" references with this change?

guybedford replied:
Yes these are no longer referenced as these conditions are not currently in use in those ecosystems as far as I'm aware. We could list these conditions if those environments wanted per the listing process described.

mcollina reviewed (approved):
LGTM

I'm generally ok with the current wording.

I would still see if anybody from the @nodejs/tsc object to this.

mcollina commented on doc/api/packages.md:
```diff
@@ -596,34 +603,30 @@ The `"import"`, `"require"`, `"node"`, `"node-addons"` and `"default"`
 These user conditions can be enabled in Node.js via the [`--conditions` flag][].
 
-The following condition definitions are currently endorsed by Node.js:
+Since user package conditions require clear definitions to ensure correct usage,
```
I think this could be called out more clearly with a subheading.

guybedford replied:
I've gone ahead and renamed the entire section to "community conditions definitions", along with further wording simplifications. Let me know if that works for you.

guybedford commented:
@weswigham in the TSC meeting today we discussed that it would be good to verify that there is implementer agreement with these conditions definitions. Can you perhaps explicitly confirm / approve you are happy with the `"types"` definition as written in this PR?

weswigham commented:
Yes with one caveat - it's fine for different conditions of an entrypoint to have different APIs (a la `module` vs `require`) - they just need separate `types` entries then; which is one of the first times people may be exposed to nested conditions. Eg,

```json
{
  "exports": {
    ".": {
      "types": {
        "module": "./types/wrapper.d.mts",
        "default": "./types/entrypoint.d.ts"
      },
      "module": "./wrapper.mjs",
      "default": "./entrypoint.js"
    }
  }
}
```

And, in fact, you typically _must_ provide a separate definition for each version of an entrypoint in such a scenario (to at least capture the differing format data).

guybedford commented:
Thanks for the confirmation @weswigham. Point taken re the dual typing - to avoid getting too much into the details I've removed the note about singular typing interface requirement for now. If you have a resource on this that can be linked to in future for further guidance please do share it here.

GeoffreyBooth commented:
> If you have a resource on this that can be linked to in future for further guidance please do share it here.

This would perhaps be a great thing for TypeScript to document, and we could link to that.

guybedford commented:
This PR no longer has a block and as per the TSC discussion we have now had implementer feedback from both Deno and TypeScript on these conditions as well. Does anyone feel any further requirements should be met to land this PR? If not I will aim to land in the coming days.

mhdawson reviewed (approved):
LGTM

ljharb commented:
@guybedford can you link to that implementer feedback?

guybedford commented:
@ljharb see https://github.com/nodejs/node/pull/40708#discussion_r752172964 and https://github.com/nodejs/node/pull/40708#issuecomment-971830856.

guybedford commented:
Landed in https://github.com/nodejs/node/commit/e31d1cb55da6ad9d05ed98d22164c01501ca2b39.
